### PR TITLE
Add Kinesis Source

### PR DESCRIPTION
**Pull request by zsxwing**

### CODE REVIEW

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+      parameters: Map[String, String]): Source = {
+    val caseInsensitiveOptions = new CaseInsensitiveMap(parameters)
+    val regionName = caseInsensitiveOptions.getOrElse("regionName", {
+      throw new IllegalArgumentException("Option 'regionName' is not specified")
+    })
+    val endpointUrl = caseInsensitiveOptions.getOrElse("endpointUrl", {
+      throw new IllegalArgumentException("Option 'endpointUrl' is not specified")
+    })
+    val streamNames = caseInsensitiveOptions.getOrElse("streamNames", {
+      throw new IllegalArgumentException("Option 'streamNames' is not specified")
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: this is the most important option, and needs to be named first.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+import org.apache.spark.{Logging, SparkContext, SparkEnv, SparkException}
+import org.apache.spark.network.util.JavaUtils
+import org.apache.spark.storage.{BlockId, StorageLevel}
+
+/**
+ * However, this class runs in the driver so could be a bottleneck.
+ */
+private[kinesis] class KinesisDataFetcher(
+  credentials: SerializableAWSCredentials,
+  endpointUrl: String,
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala]: Only endpoint is sufficient. Region name should not be needed.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  def fetch(sc: SparkContext): Array[(BlockId, SequenceNumberRange)] = {
+    sc.makeRDD(fromSeqNums, fromSeqNums.size).map {
+      case (shard, fromSeqNum) => fetchPartition(shard, fromSeqNum)
+    }.collect().flatten
+  }
+
+  private def fetchPartition(
+    shard: Shard,
+    fromSeqNum: Option[String]): Option[(BlockId, SequenceNumberRange)] = {
+    client.setEndpoint(endpointUrl, "kinesis", regionName)
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala]: This should be `client.setEndpoint(endpointUrl)`

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  override def shortName(): String = "kinesis"
+
+  override def createSource(
+      sqlContext: SQLContext,
+      schema: Option[StructType],
+      providerName: String,
+      parameters: Map[String, String]): Source = {
+    val caseInsensitiveOptions = new CaseInsensitiveMap(parameters)
+    val regionName = caseInsensitiveOptions.getOrElse("regionName", {
+      throw new IllegalArgumentException("Option 'regionName' is not specified")
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: We should not check for both regionName and endpointUrl. Only one of them is sufficient. Ultimately the source should take only endpointUrl. If the user specifies region, then you can derive the endpoint from it. 
`  Regions.getCurrentRegion.getServiceEndpoint("kinesis")`

Also make the names "region" and "endpoint".

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala (hunk 2 above)
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala]: Nvm. Other classes used in this Source needs both. So this is fine.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    }).split(',').toSet
+    val initialPosInStream =
+      caseInsensitiveOptions.getOrElse("initialPosInStream", "LATEST") match {
+        case "LATEST" => InitialPositionInStream.LATEST
+        case "TRIM_HORIZON" => InitialPositionInStream.TRIM_HORIZON
+        case pos =>
+          throw new IllegalArgumentException(s"Unknown value of option initialPosInStream: $pos")
+      }
+
+    if ((caseInsensitiveOptions.get("AWS_ACCESS_KEY_ID").nonEmpty
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: I think its weird to have the options as caps. Better to simply use "accessKey" and "secretKey".

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+import org.apache.spark.{Logging, SparkContext, SparkEnv, SparkException}
+import org.apache.spark.network.util.JavaUtils
+import org.apache.spark.storage.{BlockId, StorageLevel}
+
+/**
+ * However, this class runs in the driver so could be a bottleneck.
+ */
+private[kinesis] class KinesisDataFetcher(
+  credentials: SerializableAWSCredentials,
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala]: incorrect indentation.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala [hunk 7]
```diff
@@ -86,21 +72,36 @@ private[kinesis] case class KinesisSource(
...
   client.setEndpoint(endpointUrl, "kinesis", regionName)
 
-  override def schema: StructType = encoder.schema
+  private var cachedBlocks = new mutable.HashSet[BlockId]
+
+  override val schema: StructType = encoder.schema
 
   override def getNextBatch(start: Option[Offset]): Option[Batch] = {
-    val startOffset = start.map(KinesisSourceOffset.fromOffset)
+    val startOffset = start.map(_.asInstanceOf[KinesisSourceOffset])
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala]: This can be moved below the comparison `return None`

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisTestUtils.scala [hunk 8]
```diff
@@ -54,6 +54,8 @@ private[kinesis] class KinesisTestUtils extends Logging {
   @volatile
   private var _streamName: String = _
 
+  private val shadeIdToLatestSeqNum = mutable.HashMap[String, String]()
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisTestUtils.scala]: `shade` -> `shard`

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/package.scala [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+
+import org.apache.spark.sql.DataFrameReader
+
+package object kinesis {
+
+  /**
+   * Add the `kinesis` method to DataFrameReader that allows people to read from Kinesis using
+   * the DataFileReader.
+   */
+  implicit class KinesisDataFrameReader(reader: DataFrameReader) {
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/package.scala]: Is this pattern used in other data sources? This is cool for scala, but does not work for Java and so if we use this in examples, it would be confusing for Java users.

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala [hunk 10]
```diff
@@ -17,63 +17,158 @@
...
-class KinesisSourceSuite extends StreamTest with SharedSQLContext with KinesisFunSuite {
-
-  import testImplicits._
-
-  private var testUtils: KPLBasedKinesisTestUtils = _
-  private var streamName: String = _
-
-  override val streamingTimout = 60.seconds
+  case class AddKinesisData(
+      testUtils: KPLBasedKinesisTestUtils,
```
Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: Why pass the testUtils?

Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: Got it, nvm.

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala [hunk 11]
```diff
@@ -17,63 +17,158 @@
...
+        Assert(assertStreamBlocks, "Old stream blocks should be cleaned")
+      )
+    } finally {
+      testUtils.deleteStream()
+    }
   }
 
-  override def afterAll(): Unit = {
-    if (testUtils != null) {
+  testIfEnabled("failover") {
```
Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: failover anyways tests receiving. So just merge these two unit tests.

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala [hunk 12]
```diff
@@ -17,63 +17,158 @@
...
-    val testData = 1 to 10 // This ensures that data is sent to multiple shards for 2 shard streams
-    testStream(mapped)(
-      AddKinesisData(kinesisSource, testData: _*),
-      CheckAnswer(testData.map { _ + 1 }: _*)
-    )
+  testIfEnabled("DataFrameReader") {
+    val testUtils = new KPLBasedKinesisTestUtils
+    testUtils.createStream()
+    try {
+      val df = sqlContext.read
```
Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: Test the different possible failure conditions as well.

File: sql/core/src/test/scala/org/apache/spark/sql/StreamTest.scala [hunk 13]
```diff
@@ -71,7 +71,7 @@ trait StreamTest extends QueryTest with Timeouts {
   }
 
   /** How long to wait for an active stream to catch up when checking a result. */
-  val streamingTimeout = 10.seconds
+  def streamingTimeout: Span = 10.seconds
```
Comment (tdas) [inline sql/core/src/test/scala/org/apache/spark/sql/StreamTest.scala]: why this change?

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+    val caseInsensitiveOptions = new CaseInsensitiveMap(parameters)
+    val regionName = caseInsensitiveOptions.getOrElse("regionName", {
+      throw new IllegalArgumentException("Option 'regionName' is not specified")
+    })
+    val endpointUrl = caseInsensitiveOptions.getOrElse("endpointUrl", {
+      throw new IllegalArgumentException("Option 'endpointUrl' is not specified")
+    })
+    val streamNames = caseInsensitiveOptions.getOrElse("streamNames", {
+      throw new IllegalArgumentException("Option 'streamNames' is not specified")
+    }).split(',').toSet
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: eliminate empty names.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala (hunk 1 above)
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: Add an example in the error message to show how names can be comma separated.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala [hunk 15]
```diff
@@ -125,189 +130,50 @@ private[kinesis] case class KinesisSource(
...
-        logWarning(s"Error fetching data from shard $shard", e)
-        None
+      case e: AbortedException =>
+        // AbortedException will be thrown if the current thread is interrupted
+        // So let's convert it back to InterruptedException
+        val e1 = new InterruptedException("thread is interrupted")
+        e1.addSuppressed(e)
+        throw e1
     }
   }
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala]: Add a toString method to describe the source.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    if ((caseInsensitiveOptions.get("AWS_ACCESS_KEY_ID").nonEmpty
+      && caseInsensitiveOptions.get("AWS_SECRET_ACCESS_KEY").isEmpty)) {
+      throw new IllegalArgumentException(
+        s"AWS_ACCESS_KEY_ID is set but AWS_SECRET_ACCESS_KEY is not found")
+    }
+    if ((caseInsensitiveOptions.get("AWS_ACCESS_KEY_ID").isEmpty
+      && caseInsensitiveOptions.get("AWS_SECRET_ACCESS_KEY").nonEmpty)) {
+      throw new IllegalArgumentException(
+        s"AWS_SECRET_ACCESS_KEY is set but AWS_ACCESS_KEY_ID is not found")
+    }
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: Check whether access keys are available (either specified, or loaded from a provider) right here. Otherwise, it fails later, and gives a bad error message like 
```
com.amazonaws.AmazonClientException: Unable to load AWS credentials from any provider in the chain
	at com.amazonaws.auth.AWSCredentialsProviderChain.getCredentials(AWSCredentialsProviderChain.java:117)
	at org.apache.spark.streaming.kinesis.KinesisSource$$anonfun$5.apply(KinesisSource.scala:85)
```
So better to resolve the credentials right here and pass it on to the source without an option.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/package.scala (hunk 9 above)
Comment (zsxwing) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/package.scala]: > Is this pattern used in other data sources? This is cool for scala, but does not work for Java and so if we use this in examples, it would be confusing for Java users.

Similar to https://github.com/databricks/spark-avro

Scala usage: `val df = sqlContext.read.avro("src/test/resources/episodes.avro")`
Java usage: `DataFrame df = sqlContext.read().format("com.databricks.spark.avro").load("src/test/resources/episodes.avro");`

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+      providerName: String,
+      parameters: Map[String, String]): Source = {
+    val caseInsensitiveOptions = new CaseInsensitiveMap(parameters)
+
+    val streams = caseInsensitiveOptions.getOrElse("streams", {
+      throw new IllegalArgumentException("Option 'streams' is not specified")
+    }).split(",", -1).toSet
+    if (streams.isEmpty || streams.exists(_.isEmpty)) {
+      throw new IllegalArgumentException(
+        "Option 'streams' is invalid. Please use comma separated string (e.g., 'stream1,stream2')")
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: Option 'streams' has empty names.
BTW, isnt using `require more concise?

Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: Actually can you add both `stream` and `streams`. Either one needs to be specified.
So something like 
```
caseInsensitiveOptions.get("stream")
       .orElse(caseInsensitiveOptions.get("streams"))
       .getOrElse {   //exception }
```

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+
+  override def createSource(
+      sqlContext: SQLContext,
+      schema: Option[StructType],
+      providerName: String,
+      parameters: Map[String, String]): Source = {
+    val caseInsensitiveOptions = new CaseInsensitiveMap(parameters)
+
+    val streams = caseInsensitiveOptions.getOrElse("streams", {
+      throw new IllegalArgumentException("Option 'streams' is not specified")
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: Option `stream` needs to be specified.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+          throw new IllegalArgumentException(
+            s"'region'(${_region}) doesn't match to 'endpoint'(${_endpoint})")
+        }
+        (_region, _endpoint)
+      case (Some(_region), None) =>
+        (_region, RegionUtils.getRegion(_region).getServiceEndpoint("kinesis"))
+      case (None, Some(_endpoint)) =>
+        (RegionUtils.getRegionByEndpoint(_endpoint).getName, _endpoint)
+      case (None, None) =>
+        throw new IllegalArgumentException("Either 'region' or 'endpoint' should be specified")
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: Either option 'region' or option 'endpoint' must be specified. Examples: ...

(see example above)

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+    initialPositionInStream: InitialPositionInStream,
+    readTimeoutMs: Long = 2000L
+) extends Serializable with Logging {
+
+  /**
+   * Use lazy because the client needs to be created in executors
+   */
+  @transient private lazy val client = new AmazonKinesisClient(credentials)
+
+  def fetch(sc: SparkContext): Array[(BlockId, SequenceNumberRange)] = {
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala]: Add scala docs.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+    sc.makeRDD(fromSeqNums, fromSeqNums.size).map {
+      case (shard, fromSeqNum) => fetchPartition(shard, fromSeqNum)
+    }.collect().flatten
+  }
+
+  private def fetchPartition(
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala]: Add scala docs

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+      case (Some(accessKey), Some(secretKey)) =>
+        new SerializableAWSCredentials(accessKey, secretKey)
+      case (Some(accessKey), None) =>
+        throw new IllegalArgumentException(
+          s"'accessKey' is set but 'secretKey' is not found")
+      case (None, Some(secretKey)) =>
+        throw new IllegalArgumentException(
+          s"'secretKey' is set but 'accessKey' is not found")
+      case (None, None) =>
+        SerializableAWSCredentials(new DefaultAWSCredentialsProviderChain().getCredentials())
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: If credential is not found through provider, the error should be something like 
```
No credential found using default AWS provider chain. Specify credentials using options 'accessKey' and 'secretKey'.
```

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala [hunk 23]
```diff
@@ -125,189 +123,57 @@ private[kinesis] case class KinesisSource(
...
 
+  override def toString: String = s"KinesisSource[" +
+    s"regionName=$regionName, " +
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisSource.scala]: this is too much. just having the stream names are sufficient.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+      case (Some(_region), None) =>
+        (_region, RegionUtils.getRegion(_region).getServiceEndpoint("kinesis"))
+      case (None, Some(_endpoint)) =>
+        (RegionUtils.getRegionByEndpoint(_endpoint).getName, _endpoint)
+      case (None, None) =>
+        throw new IllegalArgumentException("Either 'region' or 'endpoint' should be specified")
+    }
+
+    val initialPosInStream =
+      caseInsensitiveOptions.getOrElse("initialPosInStream", "LATEST") match {
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: initialPosInStream is too long. just position is sufficient.

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+        case "LATEST" => InitialPositionInStream.LATEST
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/DefaultSource.scala]: this should be case agnostic.

**Comment by zsxwing** [comment]
Updated again.

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala [hunk 26]
```diff
@@ -17,63 +17,215 @@
...
-      Set(streamName),
-      InitialPositionInStream.TRIM_HORIZON)
-    val mapped = kinesisSource.toDS().map[Int]((bytes: Array[Byte]) => new String(bytes).toInt + 1)
-    val testData = 1 to 10 // This ensures that data is sent to multiple shards for 2 shard streams
-    testStream(mapped)(
-      AddKinesisData(kinesisSource, testData: _*),
-      CheckAnswer(testData.map { _ + 1 }: _*)
-    )
+  testIfEnabled("call kinesis when not using stream") {
+    val expectedMessage = "org.apache.spark.streaming.kinesis.DefaultSource is " +
```
Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: since this message is generated by some other component other than kinesis, if they change message this function could silently break (as kinesis tests dont run always). So dont check the message.

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala [hunk 27]
```diff
@@ -17,63 +17,215 @@
...
+        .kinesis().stream()
+
+      val sources = df.queryExecution.analyzed.collect {
+          case StreamingRelation(s: KinesisSource, _) => s
+        }
+      assert(sources.size === 1)
+
+      // stream
+      assertExceptionAndMessage[IllegalArgumentException](
+        "Option 'stream' must be specified. Examples: " +
```
Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: I dont like exception message testing to be so specific, but I dont have strong feelings. I usually check for the keywords being present .. like the stream name.

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala [hunk 28]
```diff
@@ -17,63 +17,215 @@
...
+    } finally {
+      testUtils.deleteStream()
+    }
   }
 
-  override def afterAll(): Unit = {
-    if (testUtils != null) {
+  testIfEnabled("DataFrameReader") {
+    val testUtils = new KPLBasedKinesisTestUtils
+    testUtils.createStream()
```
Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: why create a new stream. why not share the stream with the previous test. saves a lot of time.

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala [hunk 29]
```diff
@@ -17,63 +17,215 @@
...
+        .option("position", InitialPositionInStream.TRIM_HORIZON.name())
+        .kinesis().stream()
+
+      val sources = df.queryExecution.analyzed.collect {
+          case StreamingRelation(s: KinesisSource, _) => s
+        }
+      assert(sources.size === 1)
+
+      // stream
+      assertExceptionAndMessage[IllegalArgumentException](
```
Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: the failure checks can run even if ENABLE_KINESIS_TESTS is not set, as these tests really should not require an actual stream.

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala [hunk 30]
```diff
@@ -17,63 +17,215 @@
...
-      testUtils.endpointUrl,
-      Set(streamName),
-      InitialPositionInStream.TRIM_HORIZON)
-    val mapped = kinesisSource.toDS().map[Int]((bytes: Array[Byte]) => new String(bytes).toInt + 1)
-    val testData = 1 to 10 // This ensures that data is sent to multiple shards for 2 shard streams
-    testStream(mapped)(
-      AddKinesisData(kinesisSource, testData: _*),
-      CheckAnswer(testData.map { _ + 1 }: _*)
-    )
+  testIfEnabled("call kinesis when not using stream") {
```
Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: this test can be tested even if kinesis tests are not enabled.

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala [hunk 31]
```diff
@@ -17,63 +17,209 @@
...
+    val df = sqlContext.read
+      .option("endpoint", KinesisTestUtils.endpointUrl)
+      .option("stream", "stream1")
+      .option("accessKey", "accessKey")
+      .option("secretKey", "secretKey")
+      .option("position", InitialPositionInStream.TRIM_HORIZON.name())
+      .kinesis().stream()
+
+    val sources = df.queryExecution.analyzed.collect {
+        case StreamingRelation(s: KinesisSource, _) => s
```
Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: incorrect indent.

File: extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala [hunk 32]
```diff
@@ -17,63 +17,209 @@
...
+class KinesisSourceSuite extends KinesisSourceTest with KinesisFunSuite {
+
+  import testImplicits._
+
+  testIfEnabled("basic receiving and failover") {
+    var streamBlocksInLastBatch: Seq[StreamBlockId] = Seq.empty
 
-  override def afterAll(): Unit = {
-    if (testUtils != null) {
+    def assertStreamBlocks: Boolean = {
```
Comment (tdas) [inline extras/kinesis-asl/src/test/scala/org/apache/spark/streaming/kinesis/KinesisSourceSuite.scala]: rename to assertBlocksCleaner, and explain what it does

File: extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala [hunk 33]
```diff
@@ -0,0 +1,204 @@
...
+/**
+ * However, this class runs in the driver so could be a bottleneck.
+ */
+private[kinesis] class KinesisDataFetcher(
+    credentials: SerializableAWSCredentials,
+    endpointUrl: String,
+    fromSeqNums: Seq[(Shard, Option[String], BlockId)],
+    initialPositionInStream: InitialPositionInStream,
+    readTimeoutMs: Long = 2000L
+) extends Serializable with Logging {
```
Comment (tdas) [inline extras/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDataFetcher.scala]: incorrect indent
